src/Language/Rust.ts

```diff
@@ -0,0 +1,306 @@
+            unionType => {
+                const nullable = nullableFromUnion(unionType);
+
+                if (nullable) return this.nullableRustType(nullable, withIssues);
+
+                const [hasNull,] = removeNullFromUnion(unionType);
+
+                const isCycleBreaker = this.isCycleBreakerType(unionType);
+
+                const name = isCycleBreaker ? ["Box<", this.nameForNamedType(unionType), ">"] : this.nameForNamedType(unionType);
+
+                return hasNull
+                    ? ["Option<", name, ">"] as Sourcelike
+                    : name;
+            }
+        );
+    };
+
+    private breakCycle = (t: Type, withIssues: boolean): any => {
+        const rustType = this.rustType(t, withIssues);
+        const isCycleBreaker = this.isCycleBreakerType(t);
+
+        return isCycleBreaker ? ["Box<", rustType, ">"] : rustType;
+    }
+
+    private emitStructDefinition = (c: ClassType, className: Name): void => {
+        this.emitLine("#[derive(Serialize, Deserialize)]");
+
+        const structBody = () => this.forEachClassProperty(c, "none", (name, jsonName, prop) => {
+
+            const escapedName = rustStringEscape(jsonName);
+
+            this.emitLine("#[serde(rename = \"", escapedName, "\")]");
+
+            this.emitLine(name, ": ", this.breakCycle(prop.type, true), ",");
+        });
+
+        this.emitBlock(["pub struct ", className], structBody);
+    };
+
+    private emitBlock = (line: Sourcelike, f: () => void): void => {
+        this.emitLine(line, " {");
+        this.indent(f);
+        this.emitLine("}");
+    };
+
+    private emitUnion = (u: UnionType, unionName: Name): void => {
+        const isMaybeWithSingleType = nullableFromUnion(u);
+
+        if (isMaybeWithSingleType !== null) {
+            return;
+        }
+
+        this.emitLine("#[derive(Serialize, Deserialize)]");
+        this.emitLine("#[serde(untagged)]");
+
+        const [, nonNulls] = removeNullFromUnion(u);
+
+        this.emitBlock(["pub enum ", unionName], () => this.forEachUnionMember(u, nonNulls, "none", null,
+            (fieldName, t) => {
+                const rustType = this.breakCycle(t, true);
+                this.emitLine([fieldName, "(", rustType, "),"]);
+            }));
+    }
+
+    emitEnumDefinition = (e: EnumType, enumName: Name): void => {
+        this.emitLine("#[derive(Serialize, Deserialize)]");
+
+        this.emitBlock(["pub enum ", enumName], () => this.forEachEnumCase(e, "none",
+            (name, jsonName) => {
+                const escapedName = rustStringEscape(jsonName);
+
+                this.emitLine("#[serde(rename = \"", escapedName, "\")]");
+                this.emitLine([name, ","]);
+            }));
+    };
+
+    emitTopLevelAlias = (t: Type, name: Name): void => {
+        this.emitLine(
+            "pub type ",
+            name,
+            " = ",
+            this.rustType(t),
+            ";"
+        );
+    };
+
+    protected emitSourceStructure(): void {

```

Comment:
Please generate comments at the top of the output that explain how to read models from JSON.

Reply:
Added usage comments.